src/task.c

```diff
@@ -632,109 +627,53 @@ task_enqueue(task_system *ts,
       tq->head = tq->tail = new_task;
    }
 
-   __sync_fetch_and_add(&group->current_outstanding_tasks, 1);
+   __sync_fetch_and_add(&group->current_waiting_tasks, 1);
 
    if (group->use_stats) {
       new_task->enqueue_time = platform_get_timestamp();
    }
-   if (group->current_outstanding_tasks > group->max_outstanding_tasks) {
-      group->max_outstanding_tasks = group->current_outstanding_tasks;
-   }
-   return task_unlock_task_queue(group);
-}
-
-/*
- * task_perform_all() - Perform all tasks queued with the task system.
- * Returns as soon as it finds the queue is empty.
- */
-void
-task_perform_all(task_system *ts)
-{
-   if (!ts->use_bg_threads) {
-      platform_status rc;
-      do {
-         rc = task_perform_one(ts);
-         debug_assert(SUCCESS(rc) || STATUS_IS_EQ(rc, STATUS_TIMEDOUT));
-      } while (STATUS_IS_NE(rc, STATUS_TIMEDOUT));
-
-   } else {
-      // wait for bg threads to finish running all the queued up tasks.
-      for (task_type type = TASK_TYPE_FIRST; type != NUM_TASK_TYPES; type++) {
-         task_group *group = &ts->group[type];
-         while (group->current_outstanding_tasks != 0) {
-            platform_sleep(USEC_TO_NSEC(100000)); // 100 msec.
-         }
+   if (group->use_stats) {
+      const threadid tid = platform_get_tid();
+      if (group->current_waiting_tasks
+          > group->stats[tid].max_outstanding_tasks) {
+         group->stats[tid].max_outstanding_tasks = group->current_waiting_tasks;
       }
    }
+   platform_condvar_signal(&group->cv);
+   return platform_condvar_unlock(&group->cv);
 }
 
-static void
-task_queue_unlock(void *arg)
-{
-   task_group *group = (task_group *)arg;
-   platform_assert(!task_system_use_bg_threads(group->ts));
-   platform_mutex_unlock(&group->fg.mutex);
-}
-
-static inline platform_status
-task_group_perform_one(task_group *group)
+static platform_status
+task_group_perform_one(task_group *group, uint64 min_backlog)
```

Comment:
Would be helpful to add a prolog documenting the behaviour controlled by `min_backlog` argument.